reproducibility_tests/reproducibility_utils.jl

```diff
@@ -118,57 +159,156 @@ function latest_comparable_paths(;
-            ref_counter_file = joinpath(path, "ref_counter.jl")
-            if !isfile(ref_counter_file)
-                push!(paths_to_delete, path)
+"""
+    compute_bins(root_path::String)
+    compute_bins(sorted_paths::Vector{String})
+
+Returns a vector of reproducibility bins.
+
+Each bin is sorted such that
+`bin[1]` is the newest and `bin[end]`
+is the oldest comparable reference.
+
+```
+comparable states
+         |                           ref counter changes ---->                         | oldest
+         |                                                                             |
+         |  bin 1      bin 2      bin 3      bin 4      bin 5      bin 6      bin 7    |
+         |                                                                             |
+         |  02_49f92   04_36ebe   05_beb8a   06_4d837   05_8c311   08_45875   10_bc1e0 |
+         |             04_d6e48              06_d6d73              08_1cc58            |
+         v             04_4c042                                                        v newest
+```
+"""
+compute_bins(root_path::String) =
+    compute_bins(reverse(sorted_dataset_folder(; dir = root_path)))
+function compute_bins(sorted_paths::Vector{String})
+    isempty(sorted_paths) && return Vector{String}[]
+    bins = Vector{String}[]
+    local paths_per_bin
+    p_first_next = nothing
+    k = 1
+    while true
+        paths_per_bin = String[]
+        k > length(sorted_paths) && break
+        while true
+            k > length(sorted_paths) && break
+            p = sorted_paths[k]
+            rcp = read_ref_counter(joinpath(p, "ref_counter.jl"))
+            if isempty(paths_per_bin)
+                push!(paths_per_bin, p)
+                k += 1
             else
-                ref_counter = parse(Int, first(readlines(ref_counter_file)))
-                # Just to be safe, let's also make sure that we don't delete
-                # any paths with recent (let's say 5) ref counter increments ago.
-                if ref_counter + keep_latest_ref_counters < ref_counter_main
-                    push!(paths_to_delete, path)
+                rcb = read_ref_counter(
+                    joinpath(first(paths_per_bin), "ref_counter.jl"),
+                )
+                if rcp == rcb
+                    push!(paths_per_bin, p)
+                    k += 1
+                else
+                    break
                 end
             end
         end
-        @show ref_counter_main
-        @show length(sorted_paths)
-        @show length(paths_to_delete)
-        @info "Deleting files:"
-        for i in 1:length(paths_to_delete)
-            f = paths_to_delete[i]
-            @info "     (File, date): ($(f), $(Dates.unix2datetime(stat(f).mtime))). Reason: $(reason(f))"
-        end
-        for i in 1:length(paths_to_delete)
-            rm(paths_to_delete[i]; recursive = true, force = true)
+        push!(bins, paths_per_bin)
+    end
+    return bins
+end
```

Comment:
I think the logic of this function hard to follow. Some of the the issues are: 
- there are two nested and interacting `while true`  (both of which start with an exit condition, meaning that we can just turn them into `while condition`)
- variable names are obscure
- the intent of the various instructions is not very clear to me

I tried writing it differently, but I am not sure if I am preserving the logic:
```suggestion
function compute_bins(sorted_paths::Vector{String})
    num_paths = length(sorted_paths)
    bins = Vector{String}[]
    path_index = 1
    while path_index <= num_paths
            paths_per_bin = String[]
            while path_index <= num_paths
                first_path = sorted_paths[path_index]
                first_ref_counter = read_ref_counter(joinpath(first_path, "ref_counter.jl"))
                push!(paths_per_bin, first_path)
                path_index += 1
                current_path = sorted_paths[path_index]
                current_ref_counter = read_ref_counter(joinpath(p, "ref_counter.jl"))
                if current_ref_counter == first_ref_counter
                    push!(paths_per_bin, current_path)
                    path_index += 1
                else
                    break
                end
        end
        push!(bins, paths_per_bin)
    end
    return bins
end
```
This version also doesn't make the intent very clear

Reply:
Hm, I'm getting BoundsError with this version, also, needing to change `current_ref_counter = read_ref_counter(joinpath(p, "ref_counter.jl"))` to `current_ref_counter = read_ref_counter(joinpath(current_path, "ref_counter.jl"))`

Reply:
I can try to improve the local variable names

Reply:
I've updated the names and changed the `while` conditions